framework/src/transfers/MultiAppProjectionTransfer.C

```diff
@@ -220,44 +223,96 @@ MultiAppProjectionTransfer::assembleL2To(EquationSystems & es, const std::string
-  std::vector<MeshFunction *> from_fns(n_apps, NULL);
-  std::vector<MeshTools::BoundingBox *> from_bbs(n_apps, NULL);
-
-  // get bounding box, mesh function and solution for each subapp
-  for (unsigned int i = 0; i < n_apps; i++)
+  /********************
+  In order to assemble the system we need to evaluate the sub app solutions at
+  quadrature points in the master domain.  This is complicated by the fact that
+  each app is on its own MPI communicator.
+
+  Each processor will
+  1. Generate bounding boxes and mesh functions for it's local sub apps
+  2. Send the bounding boxes to the other processors in an all-to-all comm
+  3. Check its local quadrature points in the master domain to see which
+     bounding boxes they lie in
+  4. Send quadrature points to the processors with the containing bounding boxes
+  5. Recieve quadrature points from other processors, evaluate its mesh
+     functions at those points, and send the values back to the proper processor
+  6. Recieve mesh function evaluations from all relevant processors and decide
+     which one to use at every quadrature point (the lowest global app index
+     always wins)
+  7. And assemble the L2 system at its local quadrature points
+  ********************/
+
+  /********************
+  First step, get the bounding boxes and mesh functions for all the sub apps on
+  this processor.
+  ********************/
+  const unsigned int n_global_apps = _multi_app->numGlobalApps();
+  const unsigned int n_local_apps = _multi_app->numLocalApps();
+  std::vector<NumericVector<Number> *> serialized_from_solutions(n_local_apps, NULL);
+  std::vector<MeshFunction *> local_meshfuns(n_local_apps, NULL);
+  std::vector<MeshTools::BoundingBox> local_bboxes(n_local_apps);
+  std::vector<std::pair<Point, Point> > bb_points(n_local_apps);
+
+  unsigned int i_local = 0;
+  for (unsigned int i_global = 0; i_global < n_global_apps; i_global++)
   {
-    if (!_multi_app->hasLocalApp(i))
+    if (!_multi_app->hasLocalApp(i_global))
       continue;
 
-    MPI_Comm swapped = Moose::swapLibMeshComm(_multi_app->comm());
-
-    FEProblem & from_problem = *_multi_app->appProblem(i);
+    // Get the bounding box.
+    FEProblem & from_problem = *_multi_app->appProblem(i_global);
     EquationSystems & from_es = from_problem.es();
     MeshBase & from_mesh = from_es.get_mesh();
-    MeshTools::BoundingBox * app_box = new MeshTools::BoundingBox(MeshTools::processor_bounding_box(from_mesh, from_mesh.processor_id()));
-    from_bbs[i] = app_box;
+    MeshTools::BoundingBox app_box = MeshTools::BoundingBox(MeshTools::processor_bounding_box(from_mesh, from_mesh.processor_id()));
+
+    // Cast the bounding box into a pair of points to simplify MPI
+    // communication.  Translate the bounding box to the app's position.
+    bb_points[i_local] = static_cast<std::pair<Point, Point> >(app_box);
+    bb_points[i_local].first = bb_points[i_local].first + _multi_app->position(i_global);
```

Comment:
My current plan is to use a method of `MultiAppTransfer`.  I'd prefer to have it on the transfer side rather than the multiapp side, because I'm going to want bounding boxes for the "from" domain without worrying if that "from" domain is a master or sub app.  To illustrate my point, this is what I'm using now:

```cxx
std::vector<MeshTools::BoundingBox>
MultiAppTransfer::getBboxes()
{
  std::vector<std::pair<Point, Point> > bb_points(_from_meshes.size());
  for (unsigned int i = 0; i <  _from_meshes.size(); i++)
  {
    // Get a bounding box around the mesh elements that are local to the current
    // processor.
    MeshTools::BoundingBox bbox = MeshTools::processor_bounding_box(* _from_meshes[i], processor_id());

    // Translate the bounding box.
    bbox.first += _to_positions[i];
    bbox.second += _to_positions[i];

    // Cast the bounding box into a pair of points (so it can be put through
    // MPI communication).
    bb_points[i] = static_cast<std::pair<Point, Point> >(bbox);
  }

  // Serialize the bounding box points.
  _communicator.allgather(bb_points);
  std::vector<MeshTools::BoundingBox> bboxes(bb_points.size());
  for (unsigned int i = 0; i <  bb_points.size(); i++)
    bboxes[i] = static_cast<MeshTools::BoundingBox>(bb_points[i]);

  return bboxes;
}
```

Note that I can use `_from_meshes[i]` without considering master/sub apps